Source/Particles/Resampling/GridBasedMerging.H

```diff
@@ -0,0 +1,128 @@
+/* Copyright 2024 The WarpX Community
+ *
+ * This file is part of WarpX.
+ *
+ * Authors: Roelof Groenewald (TAE Technologies)
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef WARPX_GRID_BASED_MERGING_H_
+#define WARPX_GRID_BASED_MERGING_H_
+
+#include "Resampling.H"
+
+#include "Particles/WarpXParticleContainer_fwd.H"
+
+/**
+ * \brief This class implements a particle merging scheme wherein particles
+ * are clustered in phase space and particles in the same cluster is merged
+ * into two remaining particles. The scheme conserves linear momentum and
+ * kinetic energy within each cluster.
+ */
+class GridBasedMerging: public ResamplingAlgorithm {
+public:
+
+    /**
+     * \brief Default constructor of the GridBasedMerging class.
+     */
+    GridBasedMerging () = default;
+
+    /**
+     * \brief Constructor of the GridBasedMerging class
+     *
+     * @param[in] species_name the name of the resampled species
+     */
+    GridBasedMerging (std::string species_name);
+
+    /**
+     * \brief A method that performs merging for the considered species.
+     *
+     * @param[in] pti WarpX particle iterator of the particles to resample.
+     * @param[in] lev the index of the refinement level.
+     * @param[in] pc a pointer to the particle container.
+     */
+    void operator() (WarpXParIter& pti, int lev, WarpXParticleContainer* pc) const final;
+
+    /**
+     * \brief This merging routine requires functionality to sort a GPU vector
+     * based on another GPU vector. The heap-sort functions below where obtained
+     * from https://www.geeksforgeeks.org/heap-sort/ and modified for the
+     * current purpose.
+    */
+    struct HeapSort {
```

Comment:
It'd be cool if this functionality is already available in AMReX. But I'm not aware if it is.